source/opt/debug_info_manager.h

```diff
@@ -0,0 +1,132 @@
+#define SOURCE_OPT_DEBUG_INFO_MANAGER_H_
+
+#include <unordered_map>
+
+#include "source/opt/instruction.h"
+#include "source/opt/module.h"
+
+namespace spvtools {
+namespace opt {
+namespace analysis {
+
+// A class for analyzing, managing, and creating OpenCL.DebugInfo.100 extension
+// instructions.
+class DebugInfoManager {
+ public:
+  DebugInfoManager(IRContext* context);
+  // Constructs a def-use manager from the given |module|. All internal messages
+  // will be communicated to the outside via the given message |consumer|. This
+  // instance only keeps a reference to the |consumer|, so the |consumer| should
+  // outlive this instance.
+
+  DebugInfoManager(const DebugInfoManager&) = delete;
+  DebugInfoManager(DebugInfoManager&&) = delete;
+  DebugInfoManager& operator=(const DebugInfoManager&) = delete;
+  DebugInfoManager& operator=(DebugInfoManager&&) = delete;
+
+  friend bool operator==(const DebugInfoManager&, const DebugInfoManager&);
+  friend bool operator!=(const DebugInfoManager& lhs,
+                         const DebugInfoManager& rhs) {
+    return !(lhs == rhs);
+  }
+
+  // Clones DebugDeclare or DebugValue for a local variable whose result id is
+  // |orig_var_id|. Set Variable operand of the new DebugDeclare as
+  // |new_var_id|. Return the new DebugDeclare or DebugValue. If this is the
+  // first DebugDeclare or DebugValue, keep it in |local_var_id_to_dbgdecl_|.
+  Instruction* CloneDebugDeclare(uint32_t orig_var_id, uint32_t new_var_id);
+
+  // Returns id of new DebugInlinedAt. Its Line operand is the line number
+  // of |line| if |line| is not nullptr. Otherwise, its Line operand
+  // is the line number of lexical scope of |scope|. Its Scope and Inlined
+  // operands are Scope and Inlined of |scope|. Note that this function puts
+  // the new DebugInlinedAt into the tail of debug instructions.
+  uint32_t CreateDebugInlinedAt(const Instruction* line,
+                                const DebugScope& scope);
+
+  // Creates a new DebugInfoNone instruction and returns it. In addition,
+  // insert the new DebugInfoNone instruction before the head of debug
+  // instructions.
+  Instruction* CreateDebugInfoNone();
+
+  // Returns DebugInlinedAt whose id is |dbg_inlined_at_id|. If it does not
+  // exist or it is not a DebugInlinedAt instruction, return nullptr.
+  Instruction* GetDebugInlinedAt(uint32_t dbg_inlined_at_id);
+
+  // Returns DebugFunction whose Function operand is |fn_id|. If it does not
+  // exist, return nullptr.
+  Instruction* GetDebugFunction(uint32_t fn_id) {
+    auto dbg_fn_it = fn_id_to_dbg_fn_.find(fn_id);
+    return dbg_fn_it == fn_id_to_dbg_fn_.end() ? nullptr : dbg_fn_it->second;
+  }
+
+  // Clones DebugInlinedAt whose id is |dbg_inlined_at_id|. If
+  // |dbg_inlined_at_id| is not an id of DebugInlinedAt, returns nullptr.
+  // Note that this function does not insert the new DebugInlinedAt into
+  // debug instruction list of the module.
```

Comment:
The comment should be more definitive.  They are either hard requirements for the inputs or promises about the output.  When you use a term like "note", you drawing attention to something that they should be able to see somewhere else.  There is a photo and I ask you to note the person in the red coat.  You can look at the photo to see the person in the red coat.  We do not want people looking at the implementation or anything else.  We are making a promise: "This is true."  I've probably made this myself, and I want to try to get away from it.

In this case, something like "The new instruction will not be in the debug instruction of the module yet."